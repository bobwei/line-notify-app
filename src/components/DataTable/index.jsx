import React from 'react';
import { Table } from 'react-bootstrap';
import R from 'ramda';

import styles from './index.scss';

const DataTable = ({ settings, columns, data }) => (
  <Table responsive bsClass={`table ${styles.table}`}>
    <thead>
      <tr>
        {columns.map(({ key, label }) => (
          <th key={key}>
            {label || key}
          </th>
        ))}
      </tr>
    </thead>
    <tbody>
      {data
        .map(row => (
          <tr key={R.path([settings.keyField])(row)}>
            {columns.map(({ key, cell = (r, k) => R.path(R.split('.')(k))(r) }) => (
              <td key={key}>
                {cell(row, key)}
              </td>
            ))}
          </tr>
        ),
      )}
    </tbody>
  </Table>
);

DataTable.defaultProps = {
  settings: {
    keyField: 'id',
  },
  columns: [],
  data: [],
};

DataTable.propTypes = {
  settings: React.PropTypes.shape({
    keyField: React.PropTypes.string,
  }),
  columns: React.PropTypes.arrayOf(React.PropTypes.any.isRequired),
  data: React.PropTypes.arrayOf(React.PropTypes.any),
};

export default DataTable;
