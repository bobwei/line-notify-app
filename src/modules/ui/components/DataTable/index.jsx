import React from 'react';
import { Table } from 'react-bootstrap';
import R from 'ramda';

import Cell from 'modules/ui/components/Cell';

const DataTable = ({ settings, columns, data, ...restTableProps }) => (
  <Table responsive>
    <thead>
      <tr>
        {columns.map(({ prop, label }) => (
          <th key={prop}>
            {label || prop}
          </th>
        ))}
      </tr>
    </thead>
    <tbody>
      {data
        .map((row, index, rows) => (
          <tr key={R.path([settings.keyField])(row)}>
            {columns.map(({ prop, cell = Cell, ...restColumnProps }) => (
              <td key={prop}>
                {cell({ row, prop, index, rows, ...restTableProps, ...restColumnProps })}
              </td>
            ))}
          </tr>
        ),
      )}
    </tbody>
  </Table>
);

DataTable.defaultProps = {
  settings: {
    keyField: 'id',
  },
  columns: [],
  data: [],
};

DataTable.propTypes = {
  settings: React.PropTypes.shape({
    keyField: React.PropTypes.string,
  }),
  columns: React.PropTypes.arrayOf(React.PropTypes.shape({
    prop: React.PropTypes.string.isRequired,
    label: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.element,
    ]),
    cell: React.PropTypes.func,
  })),
  data: React.PropTypes.arrayOf(React.PropTypes.any),
};

export default DataTable;
