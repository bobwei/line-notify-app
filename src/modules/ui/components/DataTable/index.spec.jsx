/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import renderer from 'react-test-renderer';
import R from 'ramda';

import DataTable from './index';
import data from './spec.json';

it('can render table correctly', () => {
  const component = renderer.create(
    <DataTable
      columns={[
        {
          prop: 'id',
        },
        {
          prop: 'name',
        },
      ]}
      data={R.pipe(
        R.path(['results_json', 'search_results']),
        R.map(R.path(['listing'])),
      )(data)}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
