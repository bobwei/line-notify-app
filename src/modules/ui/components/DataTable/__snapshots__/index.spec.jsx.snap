exports[`test can render table correctly 1`] = `
<div
  className="table-responsive">
  <table
    className="table">
    <thead>
      <tr>
        <th>
          id
        </th>
        <th>
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <span>
            1933639
          </span>
        </td>
        <td>
          <span>
            modern double room, 5 min to MRT
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            15554097
          </span>
        </td>
        <td>
          <span>
            Taipei i trip❤4mins MRT❤Xingtian temple❤homey apt
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            7217332
          </span>
        </td>
        <td>
          <span>
            only5 secand to Xingtian Temple MRT
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            11643239
          </span>
        </td>
        <td>
          <span>
            Bright, cozy and clean private room
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            12200400
          </span>
        </td>
        <td>
          <span>
            Baloo\'s home
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            5458777
          </span>
        </td>
        <td>
          <span>
            地鐵與機場都超方便的和室套房
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            5332664
          </span>
        </td>
        <td>
          <span>
            North-Europe LOFT High Ceiling Room
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            8707367
          </span>
        </td>
        <td>
          <span>
            S.H中山國小捷運站5分鐘精緻C室可小團體8人套房月租
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2789610
          </span>
        </td>
        <td>
          <span>
            Taipei MRT Studio 到站即到家 MRT1 분
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            15038033
          </span>
        </td>
        <td>
          <span>
            台北市市中心五星級小豪宅飯店式管理-捷運松江南京站就在家門口
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            15537593
          </span>
        </td>
        <td>
          <span>
            HELLO 6-persons MRT Songjiang Nanjing 5 minutes
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            4900301
          </span>
        </td>
        <td>
          <span>
            suite A6 @Xingtian Temple Station 
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            8312800
          </span>
        </td>
        <td>
          <span>
            5★ location city centre MRT apt
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            1996229
          </span>
        </td>
        <td>
          <span>
            MRT Apartment-SongjiangNanjing*TWO
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            13059703
          </span>
        </td>
        <td>
          <span>
            行天宮正對面 溫馨雅房
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            12368784
          </span>
        </td>
        <td>
          <span>
            New open!! Zhongshan Elementary school MRT-Z STAR
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            1102678
          </span>
        </td>
        <td>
          <span>
            Taipei Manor I
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            15591121
          </span>
        </td>
        <td>
          <span>
            MRT松江南京全新裝潢雙人房
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
